text/0000-route-prefix.md

```diff
@@ -0,0 +1,109 @@
+
+### `ctx.url` rewrite
+
+Many middleware are expected to run conditional checks against `ctx.url`.  It is not the concern of the middleware to know whether the web app is deployed at the root of a host, or behind a route prefix.  In order to enforce this, a core middleware can strip the route prefix from `ctx.url` prior to any of the user-land middlewares' execution.  This has the benefit of allowing middlewares to be easily re-used across applications regardless of deployment concerns.
+
+e.g.
+```
+# fusion-cli: src/route-prefix-plugin.js
+const envVarsPlugin = require('./environment-variables-plugin');
+module.exports = function() {
+  const envVars = envVarsPlugin().of();
+  return function middleware(ctx, next) {
+    const prefix = envVars.prefix;
+
+    // enhance ctx.url sans prefix
+    if(ctx.url.indexOf(prefix) === 0 /*found at index 0*/) {
+      ctx.url = ctx.url.slice(prefix.length);
+    }
+  }
+}
+```
+
+### Prefix handling for `fusion-plugin-react-router`
+
+In order to ensure that the router works as expected, we will want to pass in the prefix as part of the [configuration](https://github.com/fusionjs/fusion-plugin-react-router#router).
+
+Today, the router has the option to take in a `basepath` as a prop:
+
+```
+import {Router} from 'fusion-plugin-react-router';
+
+<Router
+  location={...}
+  basename={...}
+  context={...}
+  onRoute={...}
+>{child}</Router>
+```
+[(link)](https://github.com/fusionjs/fusion-plugin-react-router#router)
+
+As user land plugins should not be accessing environment variables directly, this will likely be supplied by a user land `{config}.js` file.  This file must remain synchronized to the configuration used in the static asset plugin and the ctx re-write.  Due to this being sensitive to errors, it may be prudent to have a build time check here to ensure synchronization.
+
+# Drawbacks
+
+This is a non-trival amount of effort to implement.  In its current form, it is also prone to user error and requires duplicated configuration.  As noted below, the alternative in the [dependency injection](https://github.com/fusionjs/rfcs/pull/1/commits/8c440259291c9774b44b9dca07cf35dbf2a5f98e?short_path=9919d4f#diff-9919d4f45f05666e94a1bcb5edd7f51a) world is potentially cleaner.  Implementing the proposal above will require a migration path towards the DI implementation down the road.
+
+# Alternatives
+
+Alternatively, the open proposal for migrating towards a purer [dependency injection](https://github.com/fusionjs/rfcs/pull/1/commits/8c440259291c9774b44b9dca07cf35dbf2a5f98e?short_path=9919d4f#diff-9919d4f45f05666e94a1bcb5edd7f51a) system for the framework may yield a cleaner solution here as well.  At a high level, the solution may involve the following concerns:
+
+* Creating a `RoutePrefixPluginConfigToken` token to use for dependency injection.
```

Comment:
For what it's worth, the naming here could be better... more than likely this configuration is a single string that other plugins also depend upon.  Something like `routePrefixStringToken` may be a more accurate name.